### DISCUSSION TITLE
Wayland scaling problem

### DISCUSSION THREAD

**Discussion by ilyss01** [qutebrowser/qutebrowser#7938] [Q&A]
**Version info**:
(https://bpa.st/FTZQ)

**Does the bug happen if you start with `--temp-basedir`?**:
Yes

**Description**
I am using Hyprland with screen scaling. qute fonts and picture rendering just doesn't look right, it's pixelated. not smooth

![image](https://github.com/qutebrowser/qutebrowser/assets/47274578/dd97f420-b175-4954-829f-cd021b7265f3)

**How to reproduce**
Hyprland config
```
monitor = eDP-1, 1920x1080, auto, 1.2
env = QUTE_QT_WRAPPER, PyQt6
env = QT_QPA_PLATFORM, wayland
env = QT_AUTO_SCREEN_SCALE_FACTOR, 1
```

qute config purged, cache deleted

**Comment by The-Compiler** [comment]
Do you have multiple screens? This sounds like a duplicate of #7857.

**Comment by ilyss01** [comment]
Only one monitor. Replaced '1.2' scaling by '1' and the problem doesn't seem to appear
![image](https://github.com/qutebrowser/qutebrowser/assets/47274578/7aa60f8d-56b3-4c5d-84a9-a5528f06d930)

**Comment by ilyss01** [comment]
Other qt apps doesn't have the same problem. Like Telegram or Kdenlive, Falkon work kinda fine
![image](https://github.com/qutebrowser/qutebrowser/assets/47274578/a593e3a8-0026-463d-946c-ac27e16341ed)

**Comment by The-Compiler** [comment]
It might be something with Qt 6. Can you try with [testbrowser_webengine.py](https://github.com/qutebrowser/qutebrowser/blob/main/scripts/testbrowser/testbrowser_webengine.py) and see how that looks?

**Comment by ilyss01** [comment]
It just looks bad
(Using PyQt6)

![image](https://github.com/qutebrowser/qutebrowser/assets/47274578/636e411c-cf61-44ee-8c15-539930aa6a41)

**Comment by The-Compiler** [comment]
I just tried to reproduce with

```sh
export QT_SCALE_FACTOR=1.2
export QT_QPA_PLATFORM=wayland
```

on Sway and it looks fine to me. Given that you see it in the testbrowser as well, this isn't a qutebrowser-specific issue, but probably something with your setup. Thus, converting this to a discussion.

**Comment by lairdchris** [comment]
I'm seeing exactly the same problem with my Wayland configuration.
If I set scaling to 1.0, everything looks good within Qutebrowser, but with 1.5 it's like in the screenshots above.

I can change it at runtime with 

    hyprctl keyword monitor eDP-1,preferred,0x0,1.0

resp.

     hyprctl keyword monitor eDP-1,preferred,0x0,1.5

and restart Qutebrowser to see the effect.

**Comment by d3vv3** [comment]
I am also getting this scaling issue. Just moved from sway to hyprland, so there may be something wrong on their side (?)

**Reply by d3vv3** [reply]
Just as a sidenote, the pixelation only occurs in the actual window content, all the tab titles and qutebrowser menus look sharp.

**Reply by 4tuw2tuv0** [reply]
Which distro. were you running Sway from, @d3vv3? And (if you remember) what version of Sway was being run?

**Reply by d3vv3** [reply]
I am on Fedora, but I do not remember the main version. Either way, I am getting it onto Hyprland now. See the top right text, running on a 4k screen:

![image](https://github.com/qutebrowser/qutebrowser/assets/43572680/75927f57-226d-41b5-8137-54037e5ae475)

**Reply by 4tuw2tuv0** [reply]
Yeah; I ask because I'm wondering if it's `wlroots` v0.17 related as it looks like Sway didn't have a commit for upgrading to v0.17 until ~4 weeks ago and released their latest version just 3 weeks ago.

It might explain why everyone who was trying it on Sway (including @The-Compiler) didn't see any issue (and why I did when I tried it in Sway just yesterday) while Hyprland (perhaps; I haven't checked) might have been using v17 for a longer period of time, already, which is why you noticed right away once you switched.

**Reply by 4tuw2tuv0** [reply]
Just a theory, though.

**Comment by Xiretza** [comment]
Getting the same on sway 1.9 with one monitor, 1.5x scaling, and both qutebrowser 3.1.0 and testbrowser.

**Comment by dbeley** [comment]
Got the same issue with hyprland on Nixos with qutebrowser 3.1.0, I managed to fix it by setting `QT_SCALE_FACTOR_ROUNDING_POLICY=RoundPreferFloor` as env variable (credits: https://old.reddit.com/r/qutebrowser/comments/15xmvaz/how_do_i_get_crisp_text_on_wayland/kpfevjn/).

**Reply by iNecas** [reply]
I've hit the same issue, after updating sway from 1.8.1 to 1.9. Qutebrowser 3.1 on sway 1.8.1 was working fine, weird fonts on sway 1.9 when scaled to 1.5.  I can confirm `QT_SCALE_FACTOR_ROUNDING_POLICY=RoundPreferFloor` fixed the issue for me. Thanks @dbeley for sharing: you saved my day :)

Other relevant changes might be wlroots 0.16.2 -> 0.17.1.

**Reply by 4tuw2tuv0** [reply]
> Other relevant changes might be wlroots 0.16.2 -> 0.17.1.

Good point, @iNecas; I'd encountered this when I upgraded Wayfire. I'd thought it was Wayfire specific, at first, as I didn't notice anything in RiverWM or SwayWM but, after realizing I wasn't using scaling in either of them, I found the same behavior upon applying scaling.

The version of Wayfire I'd been using, up until that point, was the commit right before requiring an update to `wlroots` v0.17 so it could, very well, be `wlroots` that's the issue.

And thanks, @dbeley, for `QT_SCALE_FACTOR_ROUNDING_POLICY=RoundPreferFloor`; I was trying to deal with the display but it was giving me headaches, putting up with that.

**Reply by d3vv3** [reply]
That fixes it for me too, thanks!

**Reply by adigitoleo** [reply]
Same issue, same fix. Also on sway upgrading to wlroots 0.17. The link mentions that this basically disables support for the new fractional scaling protocol, so wlroots is a likely reason (I'm guessing that is where protocols get added, but I can't find a changelog on their gitlab). Maybe a relevant QtWebEngine report can be submitted.

**Reply by jezcope** [reply]
I've been trying to figure this one out all weekend and finally arrived at the same workaround as @dbeley. It has already been reported in the Qt bug tracker and is being worked on: https://bugreports.qt.io/browse/QTBUG-113574

**Comment by 4tuw2tuv0** [comment]
Has anyone had the work-around suddenly stop working for them? I don't think anything (version number, etc.) had changed but `QT_SCALE_FACTOR_ROUNDING_POLICY` no longer prevents the choppy rendering.
